python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -454,65 +556,114 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
             return df
 
 
-@dataclasses.dataclass
 class Select(IR):
     """Produce a new dataframe selecting given expressions from an input."""
 
+    __slots__ = ("exprs", "children", "should_broadcast")
+    _non_child = ("schema", "exprs", "should_broadcast")
+    children: tuple[IR]
+
     df: IR
     """Input dataframe."""
-    expr: list[expr.NamedExpr]
+    exprs: tuple[expr.NamedExpr, ...]
     """List of expressions to evaluate to form the new dataframe."""
     should_broadcast: bool
     """Should columns be broadcast?"""
 
+    def __init__(
+        self,
+        schema: Schema,
+        exprs: Sequence[expr.NamedExpr],
+        should_broadcast: bool,  # noqa: FBT001
+        df: IR,
+    ):
+        super().__init__(schema)
+        self.exprs = tuple(exprs)
+        self.should_broadcast = should_broadcast
+        self.children = (df,)
```

Comment:
It seems like we could probably avoid a lot of boilerplate code by formalizing the `__init__` logic to do something like this in Node or IR:

```python
    def __init__(self, *args):
        num_non_child = len(self._non_child)
        children = []
        for i, arg in enumerate(args):
            try:
                setattr(self, self._non_child[i], arg)
            except IndexError:
                children.append(arg)
        self.children = tuple(children)
```

Reply:
This is a good idea, let me give it a try.

Reply:
I had a go, but I think I don't like it that much. Rationale:

- It makes object construction more expensive (around 10% in my tests), which might start getting more expensive with rewrites
- Unless you write out the signature of `__init__` anyway, you get no type-checking for constructors, which is painful.
- Relatedly, you need to read too much code to figure out argument order (and non-locally).

Reply:
>It makes object construction more expensive...
>...you get no type-checking for constructors...

Ah okay - thanks for humoring me :)